backend/modules/assistant/ito/difference/difference_agent.py

```diff
@@ -0,0 +1,133 @@
+    )
+
+
+class ContextType(BaseModel):
+    """Represents the context given and the specific category if needed."""
+
+    category: str = Field(description="Category of the context")
+    context: str = Field(description="Context to add to the query")
+
+
+"""
+
+"""
+
+
+class DifferenceAgent:
+    def __init__(
+        self,
+        diff_query_engine: DiffQueryEngine,
+        document_context: str | None = None,
+        question_generator: QuestionGenerator | None = None,
+    ):
+        self.document_context = document_context
+        self.question_generator = (
+            question_generator
+            if question_generator
+            else QuestionGenerator(document_context=document_context)
+        )
+        self.diff_query_engine = diff_query_engine
+
+        self.questions = None
+        self.generated_df = pd.DataFrame()
+
+    def generate_questions(
+        self, target_content: str, language_verification: bool = False
+    ) -> list[str] | None:
+
+        self.questions = self.question_generator.generate_questions(
+            target_content=target_content, language_verification=language_verification
+        )
+
+    def run(
+        self,
+        target_content: str | None = None,
+        language_verification: bool = False,
+        additional_context: ContextType | None = None,
+        verbose: bool = False,
+        n_retry: int = 3,
+    ):
+        if self.questions is None and target_content:
+            self.questions = self.generate_questions(
+                target_content=target_content,
+                language_verification=language_verification,
+            )
+        elif self.questions is None:
+            raise Exception(
+                f"Please provide a source path and tab name to generate questions"
+            )
+
+        print("Querying generated questions to the reference document...")
+        analysis = []
+
+        async def query_all(questions):
+            return await asyncio.gather(*[query_one(question) for question in questions])  # type: ignore
+
+        async def query_one(question):
+            response = ResponseType(name=None, detailed_answer=None, decision=None)
+
+            for i in range(n_retry):
+                try:
+                    response: ResponseType = await self.diff_query_engine.query_engine.aquery(question[:-1])  # type: ignore
+                    nodes_to_update = [response.source_nodes[int(i)] for i in response.response.used_chunks]  # type: ignore
+                    self.diff_query_engine.update_query_engine(nodes_to_update)
+
+                except Exception as e:
+                    print(e)
+                    if verbose:
+                        print(f"Error with question: {question}")
+                        print("Retry ...")
+                        if i == 2:
+                            print(
+                                f"{n_retry} repeted errors with the same question, deleting the question."
```

Comment:
I think this retry algo is broken, check an implementation of retry (with exponential backoff ) : 
https://github.com/QuivrHQ/quivr/blob/f384b769f3f43cabf3b603bfefba3b03f443489c/backend/modules/sync/utils/list_files.py#L197